src/selfie_park/src/park_server.cpp

```diff
@@ -0,0 +1,353 @@
+      drive(-PARKING_SPEED, 0);
+      if (actual_back_parking_position_.x_ < back_wall_ + max_distance_to_wall_)
+      {
+        drive(0, 0);
+        parking_state_ = going_out;
+      }
+      break;
+
+    case going_out:
+      if (state_msgs_) ROS_INFO("get_out");
+      if (leave()) parking_state_ = out;
+      break;
+
+    case out:
+      drive(PARKING_SPEED, 0);
+      selfie_msgs::parkResult result;
+      result.done = true;
+      as_.setSucceeded(result);
+      parking_state_ = not_parking;
+      break;
+  }
+}
+
+void ParkService::initParkingSpot(const geometry_msgs::Polygon &msg)
+{
+  std::vector <tf::Vector3> odom_parking_spot;
+
+  for (std::vector<geometry_msgs::Point32>::const_iterator it = msg.points.begin(); it < msg.points.end(); it++)
+  {
+    tf::Vector3 vec(it->x, it->y, 0);
+    odom_parking_spot.push_back(actual_laser_odom_position_.transform_ * vec);
+  }
+  tf::Vector3 tl = odom_parking_spot[0];
+  tf::Vector3 bl = odom_parking_spot[1];
+  tf::Vector3 br = odom_parking_spot[2];
+  tf::Vector3 tr = odom_parking_spot[3];
+
+  parking_spot_position_ = Position(bl.x(), bl.y(), atan2(br.y() - bl.y(), br.x() - bl.x()));
+  actual_parking_position_ = Position(parking_spot_position_.transform_.inverse() * actual_odom_position_.transform_);
+  actual_laser_parking_position_ = Position(actual_parking_position_, ODOM_TO_LASER);
+  std::vector <tf::Vector3> parking_parking_spot;
+  for (std::vector<geometry_msgs::Point32>::const_iterator it = msg.points.begin(); it < msg.points.end(); it++)
+  {
+    tf::Vector3 vec(it->x, it->y, 0);
+    parking_parking_spot.push_back(actual_laser_parking_position_.transform_ * vec);
+  }
+  tl = parking_parking_spot[0];
+  bl = parking_parking_spot[1];
+  br = parking_parking_spot[2];
+  tr = parking_parking_spot[3];
+
+  parking_spot_width_ = tl.y() < tr.y() ? tl.y() : tr.y();
+  middle_of_parking_spot_y_ = parking_spot_width_ / 2.0;
+  back_wall_ = tl.x() > bl.x() ? tl.x() : bl.x();
+  front_wall_ = tr.x() < br.x() ? tr.x() : br.x();
+  middle_of_parking_spot_x_ = (front_wall_ - back_wall_) / 2.0;
+  mid_y_ = middle_of_parking_spot_y_ + (leaving_target_ - middle_of_parking_spot_y_) / 2.0;
+}
+
+void ParkService::goalCB()
+{
+  selfie_msgs::parkGoal goal = *as_.acceptNewGoal();
+  initParkingSpot(goal.parking_spot);
+  parking_state_ = go_to_parking_spot;
+}
+
+void ParkService::preemptCB()
+{
+  ROS_INFO("parkService preempted");
+  parking_state_ = not_parking;
+  move_state_ = first_phase;
+  as_.setPreempted();
+}
+
+
+void ParkService::drive(float speed, float steering_angle)
+{
+  ackermann_msgs::AckermannDriveStamped msg;
+  msg.header.stamp = ros::Time::now();
+  msg.drive.speed = speed;
+  msg.drive.steering_angle = steering_angle;
+  ackermann_pub_.publish(msg);
+}
+
+
+bool ParkService::toParkingSpot()
+{
+  selfie_msgs::parkFeedback feedback;
+  feedback.distance = actual_parking_position_.y_ - middle_of_parking_spot_y_;
+  as_.publishFeedback(feedback);
```

Comment:
Bez sensu jak dla mnie ten feedback odległości. Za to brakuje ważniejszego - action_status